### PR TITLE
Use edgeDNSServer for NS name resolution

### PR DESCRIPTION
**Pull request by ytsarev**
* Use edgeDNSServer for NS name resolution with the fallback to the local nameserver
* Supporting dns related functions
* Logs to stdout during the tests
* Use localhost as default edgeDNSServer to avoid network timeout

### CODE REVIEW

File: controllers/providers/assistant/gslb.go [hunk 1]
```diff
@@ -221,25 +221,50 @@ func (r *GslbLoggerAssistant) InspectTXTThreshold(fqdn string, edgeDNSServerPort
 	return errors.NewResourceExpired(fmt.Sprintf("Can't find split brain TXT record at EdgeDNS server(%s) and record %s ", ns, fqdn))
 }
 
+func GetARecords(msg *dns.Msg) []string {
```
Comment (kuritka) [inline controllers/providers/assistant/gslb.go:224]: keep it private or consider to expose via interface `Assistant`

File: controllers/providers/assistant/gslb.go [hunk 2]
```diff
@@ -221,25 +221,50 @@ func (r *GslbLoggerAssistant) InspectTXTThreshold(fqdn string, edgeDNSServerPort
...
+	var ARecords []string
+	for _, nsA := range msg.Answer {
+		IP := strings.Split(nsA.String(), "\t")[4]
+		ARecords = append(ARecords, IP)
+	}
+	return ARecords
+}
+
+func DNSQuery(fqdn string, nameserver string, nameserverport int) (*dns.Msg, error) {
```
Comment (kuritka) [inline controllers/providers/assistant/gslb.go:233]: keep it private or consider to expose via interface `Assistant`

File: controllers/providers/assistant/gslb.go [hunk 3]
```diff
@@ -221,25 +221,50 @@ func (r *GslbLoggerAssistant) InspectTXTThreshold(fqdn string, edgeDNSServerPort
 	return errors.NewResourceExpired(fmt.Sprintf("Can't find split brain TXT record at EdgeDNS server(%s) and record %s ", ns, fqdn))
 }
 
+func GetARecords(msg *dns.Msg) []string {
+	var ARecords []string
+	for _, nsA := range msg.Answer {
+		IP := strings.Split(nsA.String(), "\t")[4]
+		ARecords = append(ARecords, IP)
```
Comment (kuritka) [inline controllers/providers/assistant/gslb.go:228]: consider this:
```suggestion
		ip := nsA.(*dns.A).A.String()
		ARecords = append(ARecords, ip)
```

File: controllers/providers/assistant/gslb.go [hunk 4]
```diff
@@ -221,25 +221,50 @@ func (r *GslbLoggerAssistant) InspectTXTThreshold(fqdn string, edgeDNSServerPort
...
+	}
+	return ARecords
+}
+
+func DNSQuery(fqdn string, nameserver string, nameserverport int) (*dns.Msg, error) {
+	dnsMsg := new(dns.Msg)
+	edgeDNSServer := fmt.Sprintf("%s:%v", nameserver, nameserverport)
+	fqdn = fmt.Sprintf("%s.", fqdn) // Convert to true FQDN with dot at the end
```
Comment (kuritka) [inline controllers/providers/assistant/gslb.go:236]: should be the argument name fqdn, when it bacomes fqdn here ^^?

**Review by kuritka — CHANGES_REQUESTED** [review]
thx for fix tests, I have some points to consider

File: controllers/providers/assistant/gslb.go (hunk 4 above)
Comment (ytsarev) [reply controllers/providers/assistant/gslb.go:236]: Not sure if I got the questions right but we need to add dot(.) to the end of the argument to satisfy miekgdns expectations

File: controllers/providers/assistant/gslb.go (hunk 3 above)
Comment (ytsarev) [reply controllers/providers/assistant/gslb.go:228]: wasn't aware it is available this way, will try it out 👍

**Comment by ytsarev** [comment]
@kuritka thanks for the careful review, amended, please double check

File: controllers/providers/assistant/gslb.go (hunk 4 above)
Comment (kuritka) [reply controllers/providers/assistant/gslb.go:236]: I mean, the argument accepting by DNSQuery is not fqdn. We are making it fqdn.  So the point is, to rename input argument to e.g. `host` ?

Comment (ytsarev) [reply controllers/providers/assistant/gslb.go:236]: ah, ok, makes sense

Comment (ytsarev) [reply controllers/providers/assistant/gslb.go]: @kuritka amended

**Review by kuritka — APPROVED** [review]

**Pull request merged**